api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,578 @@
+  const { user } = req;
+  if (!user) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT('Unreachable. User not authenticated.')
+    );
+  }
+  const isExamPrerequisitesMet = checkPrerequisites(user, true);
+
+  if (!isExamPrerequisitesMet) {
+    void reply.code(403);
+    // TODO: Consider sending unmet prerequisites
+    return reply.send(
+      ERRORS.FCC_EINVAL_EXAM_ENVIRONMENT_PREREQUISITES(
+        'User has not completed prerequisites.'
+      )
+    );
+  }
+
+  // Check user has not completed exam in last 24 hours
+  const maybeExamAttempts = await mapErr(
+    this.prisma.envExamAttempt.findMany({
+      where: {
+        userId: user.id,
+        examId: exam.id
+      }
+    })
+  );
+
+  if (maybeExamAttempts.hasError) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeExamAttempts.error))
+    );
+  }
+
+  const examAttempts = maybeExamAttempts.data;
+
+  const maybeLastAttempt = examAttempts.length
+    ? examAttempts.reduce((latest, current) =>
+        latest.startTimeInMS > current.startTimeInMS ? latest : current
+      )
+    : null;
+
+  if (maybeLastAttempt) {
+    const lastAttempt = maybeLastAttempt;
+
+    const attemptIsExpired =
+      lastAttempt.startTimeInMS + exam.config.totalTimeInMS < Date.now();
+    if (attemptIsExpired) {
+      // If exam is not submitted, use exam start time + time allocated for exam
+      const effectiveSubmissionTime =
+        lastAttempt.submissionTimeInMS ??
+        lastAttempt.startTimeInMS + exam.config.totalTimeInMS;
+      const twentyFourHoursAgo = Date.now() - 24 * 60 * 60 * 1000;
+
+      if (effectiveSubmissionTime > twentyFourHoursAgo) {
+        void reply.code(429);
+        // TODO: Consider sending last completed time
+        return reply.send(
+          ERRORS.FCC_EINVAL_EXAM_ENVIRONMENT_PREREQUISITES(
+            'User has completed exam too recently to retake.'
+          )
+        );
+      }
+    } else {
+      // Camper has started an attempt, but not submitted it, and there is still time left to complete it.
+      // This is most likely to happen if the Camper's app closes and is reopened.
+      // Send the Camper back to the exam they were working on.
+      const generated = await mapErr(
+        this.prisma.envGeneratedExam.findFirst({
+          where: {
+            id: lastAttempt.generatedExamId
+          }
+        })
+      );
+
+      if (generated.hasError) {
+        void reply.code(500);
+        return reply.send(
+          ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(generated.error))
+        );
+      }
+
+      if (generated.data === null) {
+        void reply.code(500);
+        return reply.send(
+          ERRORS.FCC_ERR_EXAM_ENVIRONMENT(
+            'Unreachable. Generated exam not found.'
+          )
+        );
+      }
+
+      const userExam = constructUserExam(generated.data, exam);
```

Comment:
```suggestion
      const userExam = constructUserExam(generated.data!, exam);
```
Makes the above assertion

Reply:
The above is a check - not an assertion. So, I think it would be wrong to do this.